### PR TITLE
Add a new submodule to build a client bundle.

### PR DESCRIPTION
**Pull request by carlosalberto**
## Have a bundled Client artifact (fat jar)

Addresses #599 - I'm wondering if this name would go well with the rest of the artifacts ;)

Let me know if this could be of help.

### CODE REVIEW

File: jaeger-client/build.gradle [hunk 1]
```diff
@@ -7,3 +9,28 @@ dependencies {
 
     testCompile group: 'junit', name: 'junit', version: junitVersion
 }
+
+/* Shaded the same libraries as jaeger-thrift,
+ * leaving the slf4j dependency in place, as a *requirement*
```
Comment (objectiser) [inline jaeger-client/build.gradle]: This comment seems out of date now.

File: jaeger-client/build.gradle [hunk 2]
```diff
@@ -7,3 +9,28 @@ dependencies {
...
+ * for a standalone fat jar. */
+shadowJar {
+    relocate 'com.google.gson'  , 'io.jaegertracing.vendor.com.google.gson'
+    relocate 'com.twitter'      , 'io.jaegertracing.vendor.com.twitter'
+    relocate 'okhttp'           , 'io.jaegertracing.vendor.okhttp'
+    relocate 'okio'             , 'io.jaegertracing.vendor.okio'
+    relocate 'org.apache'       , 'io.jaegertracing.vendor.org.apache'
+    relocate 'org.slf4j'        , 'io.jaegertracing.vendor.org.slf4j'
+    classifier 'bundle'
```
Comment (objectiser) [inline jaeger-client/build.gradle]: Would prefer another name - `bundle` makes me think OSGi :smile: , but not sure what might be better. Could be something simple like `all` - but will let others comment.

File: jaeger-client/build.gradle (hunk 1 above)
Comment (objectiser) [reply jaeger-client/build.gradle]: Although not sure whether the slf4j-api should be shaded, as there is no slf4j logging impl as part of this uber jar, so would need to pick it up from the app's dependencies. Not sure what the impact of shading the api would have on picking up an impl?

File: jaeger-client/build.gradle [hunk 3]
```diff
@@ -7,3 +9,28 @@ dependencies {
...
+    }
+}
+
+artifacts {
+    archives(shadowJar.archivePath) {
+        builtBy shadowJar
+    }
+}
+
+tasks.check.dependsOn tasks.shadowJar
```
Comment (objectiser) [inline jaeger-client/build.gradle]: This wasn't required in the thrift usecase - was the shadowJar task not being triggered?

File: jaeger-client/build.gradle (hunk 1 above)
Comment (carlosalberto) [reply jaeger-client/build.gradle]: Yeah, I was wondering about this myself later on (whether shading it or not). So I think I will update the code to *not* shade it at all.

File: jaeger-client/build.gradle (hunk 2 above)
Comment (carlosalberto) [reply jaeger-client/build.gradle]: `all` sounds about right to me.

File: jaeger-client/build.gradle (hunk 3 above)
Comment (carlosalberto) [reply jaeger-client/build.gradle]: `shadowJar` is *only* being triggered in the thrift case when `jaeger-crossdock` is being built. Based on what you wrote, it *sounds* like we should do the same here? :)

Comment (objectiser) [reply jaeger-client/build.gradle]: Not sure it should be dependent on `jaeger-crossdock` build - ideally it should just be built when the non-shaded dependency is being built.

Comment (carlosalberto) [reply jaeger-client/build.gradle]: Oh, then this line should stay, as it builds the shaded artifact right after the non-shaded one was built.

**Comment by carlosalberto** [comment]
hey @objectiser @yurishkuro 

I've updated the PR to reflect the latest feedback. Let me know ;)

**Comment by objectiser** [comment]
@carlosalberto looks good - my only concern is that I believe the qualified jar dependency would still have a transient dependency on the non-shaded dependencies.

If the main use of the jar is to physically download the jar and include it with a javaagent, then this would not be an issue. But if someone includes the dependency in their app's pom.xml, this would still pull in the transient dependencies. I guess they could always exclude them - but not ideal.

What do others think?

**Comment by carlosalberto** [comment]
Hey hey @objectiser 

Trying to re-take this conversation:

> If the main use of the jar is to physically download the jar and include it with a javaagent, then this would not be an issue.

I think this is the exact use scenario, yes - to download and use it directly (would also work when doing Cassandra server side instrumentation, where you would simply put this JAR in the Cassandra bin path).

**Comment by objectiser** [comment]
@carlosalberto Agree it meets that use case - just worried whether some users would expect to use it as a pom dependency.

@jpkrohling Thoughts? Would documenting this in the readme be enough?

**Comment by jpkrohling** [comment]
@carlosalberto would you please list the contents of your local Maven repository for the affected artifact like, `tree ~/.m2/repository/io/jaegertracing/jaeger-client/`? 

If you are producing both a "slim" and a "fat" jar, I'd be interested in seeing the `mvn dependency:tree` from an application consuming each of them. If you haven't tested that yet, I can do it myself. I think I still have something I used for #526 and related.

**Pull request closed without merging**